src/Microsoft.SqlTools.ServiceLayer/Metadata/SmoMetadataFactory.cs

```diff
@@ -0,0 +1,122 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Data.SqlClient;
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.Smo;
+using Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection;
+using Microsoft.SqlTools.ServiceLayer.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Metadata
+{
+    /// <summary>
+    /// Interface for a factory that generates metadata for an object to edit
+    /// </summary>
+    public interface IMetadataFactory
+    {
+        /// <summary>
+        /// Generates a edit-ready metadata object
+        /// </summary>
+        /// <param name="connection">Connection to use for getting metadata</param>
+        /// <param name="objectName">Name of the object to return metadata for</param>
+        /// <param name="objectType">Type of the object to return metadata for</param>
+        /// <returns>Metadata about the object requested</returns>
+        TableMetadata GetObjectMetadata(DbConnection connection, string schemaName, string objectName, string objectType);
+    }
+
+    /// <summary>
+    /// Factory that generates metadata using a combination of SMO and SqlClient metadata
+    /// </summary>
+    public class SmoMetadataFactory : IMetadataFactory
```

Comment:
I think ideally, the MetadataService should expose a method for retrieving table metadata. This way the EditDataService doesn't need to take a hard dependency on a class provided inside another service's namespace. Or to put it another way, the functionality that a service needs should be discoverable via MEF. Since the MEF objects are all services, in order to stick with precedent, we'd want the functionality to be accessible via the service.

The QueryExecutionService provides a decent example of what I'm suggesting. The base functionality of executing a query and disposing of a query is provided via InterService* methods. When a different service needs access to query handling (ie, EditDataService) it calls the InterServiceExecuteQuery method on the QueryExecutionService. If a JSON RPC request to execute a query is received, it's handled by HandleExecuteQueryRequest which calls InterServiceExecuteQuery.

I'm guessing that a change like I'm suggesting is further in the pipeline, but I want to make sure we're on the same page for providing access to metadata to multiple services in the tools service.